### PR TITLE
DataGrid / TreeList - Return promise from addRow method

### PR DESCRIPTION
**Pull request by igormalzew**

### CODE REVIEW

File: js/ui/grid_core/ui.grid_core.editing.js [hunk 1]
```diff
@@ -848,27 +851,30 @@ var EditingController = modules.ViewController.inherit((function() {
...
             }
 
             if(editMode === EDIT_MODE_CELL && that.hasChanges()) {
                 that.saveEditData().done(function() {
                     // T804894
                     if(!that.hasChanges()) {
-                        that.addRow(parentKey);
+                        that.addRow(parentKey).done(() => {
+                            deferred.resolve();
+                        });
```
Comment (vconst) [inline js/ui/grid_core/ui.grid_core.editing.js]: Is there fail handler?

File: js/ui/grid_core/ui.grid_core.editing.js [hunk 2]
```diff
@@ -848,27 +851,30 @@ var EditingController = modules.ViewController.inherit((function() {
...
                 param = { data: {} },
                 editMode = getEditMode(that),
-                oldEditRowIndex = that._getVisibleEditRowIndex();
+                oldEditRowIndex = that._getVisibleEditRowIndex(),
+                deferred = new Deferred();
 
             if(!store) {
                 dataController.fireError('E1052', this.component.NAME);
-                return;
+                return deferred.reject();
```
Comment (vconst) [inline js/ui/grid_core/ui.grid_core.editing.js]: Pass error E1052 to reject

File: js/ui/grid_core/ui.grid_core.editing.js [hunk 3]
```diff
@@ -848,27 +851,30 @@ var EditingController = modules.ViewController.inherit((function() {
...
                 });
-                return;
+                return deferred.promise();
             }
 
             that.refresh();
 
             if(!that._allowRowAdding()) {
-                return;
+                return deferred.reject();
```
Comment (vconst) [inline js/ui/grid_core/ui.grid_core.editing.js]: Add "cancel" arg to reject call

File: js/ui/grid_core/ui.grid_core.editing.js [hunk 4]
```diff
@@ -879,7 +885,10 @@ var EditingController = modules.ViewController.inherit((function() {
                 if(that._allowRowAdding()) {
```
Comment (vconst) [inline js/ui/grid_core/ui.grid_core.editing.js:883]: if `that._allowRowAdding()` is false we should call `reject("cancel")`

File: js/ui/grid_core/ui.grid_core.editing.js [hunk 5]
```diff
@@ -879,7 +885,10 @@ var EditingController = modules.ViewController.inherit((function() {
...
                     that._addRowCore(param.data, parentKey, oldEditRowIndex);
                 }
+                deferred.resolve();
             });
```
Comment (vconst) [inline js/ui/grid_core/ui.grid_core.editing.js]: Is there fail handler?

File: testing/tests/DevExpress.ui.widgets.dataGrid/editing.tests.js [hunk 6]
```diff
@@ -3200,6 +3200,83 @@ QUnit.test('Remove the inserted row with edit mode batch and hidden column', fun
...
+    let promiseInitNewRowCallCount = 0;
+    let initNewRowCallCount = 0;
+
+    that.options.editing = {
+        mode: 'cell',
+        allowAdding: true
+    };
+
+    that.options.onInitNewRow = (e) => {
+        if((promiseInitNewRowCallCount + initNewRowCallCount) % 2 === 0) {
```
Comment (vconst) [inline testing/tests/DevExpress.ui.widgets.dataGrid/editing.tests.js]: Why does such complex condition use?

Comment (vconst) [reply testing/tests/DevExpress.ui.widgets.dataGrid/editing.tests.js]: You can define deferred outside onInitNewRow and resolve it manually if you want test that addRow promise is resolved after resolving e.promise in onInitNewRow

**Review by vconst — CHANGES_REQUESTED** [review]

File: js/ui/grid_core/ui.grid_core.editing.js (hunk 1 above)
Comment (igormalzew) [reply js/ui/grid_core/ui.grid_core.editing.js]: Done

File: js/ui/grid_core/ui.grid_core.editing.js (hunk 3 above)
Comment (igormalzew) [reply js/ui/grid_core/ui.grid_core.editing.js]: Done

File: js/ui/grid_core/ui.grid_core.editing.js (hunk 5 above)
Comment (igormalzew) [reply js/ui/grid_core/ui.grid_core.editing.js]: Done

File: testing/tests/DevExpress.ui.widgets.dataGrid/editing.tests.js (hunk 6 above)
Comment (igormalzew) [reply testing/tests/DevExpress.ui.widgets.dataGrid/editing.tests.js]: Done

**Review by Alyar666 — DISMISSED** [review]

**Comment by igormalzew** [comment]
https://github.com/DevExpress/DevExtreme/pull/11398

**Pull request closed without merging**